django/utils/connection.py

```diff
@@ -17,6 +17,11 @@ def __getattr__(self, item):
     def __setattr__(self, name, value):
         return setattr(self._connections[self._alias], name, value)
 
+    def __dir__(self):
+        new_dir = set(super().__dir__())
+        new_dir.update(dir(self._connections[self._alias]))
+        return new_dir
```

Comment:
`dir()` will sort and return a list -- see [`object.__dir__()`](https://docs.python.org/3/reference/datamodel.html#object.__dir__) -- but, while returning a set seems to work, the documentation for [`dir()`](https://docs.python.org/3/library/functions.html#dir) implies that a list ought to have been returned by `__dir__()`.

We can potentially simplify this:

```suggestion
        connection = self._connections[self._alias]
        return sorted(set(super().__dir__()).union(dir(connection)))
```

Reply:
I'm also wondering if this risks establishing connections when using autocomplete in a shell.

The issue is that `self._connections[self._alias]` is going to call `BaseConnectionHandler.__getitem__()` which will call `.create_connection()`. See `BaseConnectionHandler.all()` and where it is used to avoid unintentionally creating connections.

Reply:
> I'm also wondering if this risks establishing connections when using autocomplete in a shell.
> 
> The issue is that `self._connections[self._alias]` is going to call `BaseConnectionHandler.__getitem__()` which will call `.create_connection()`. See `BaseConnectionHandler.all()` and where it is used to avoid unintentionally creating connections.

That is indeed possible, since it will, eventually create a connection with any attribute. I'm wondering however if this is an issue, since eventually a connection will probably be established, although it seems to suggest it only creates the connection once (since that is probably the main idea of a proxy: postponing the creation of a connection, and if there is one, reuse it).